crates/wast/src/core/resolve/names.rs

```diff
@@ -514,25 +514,25 @@ impl<'a, 'b> ExprResolver<'a, 'b> {
                 self.resolve_block_type(&mut t.block)?;
             }
 
-            Block(bt) | If(bt) | Loop(bt) | Try(bt) => {
+            Block(bt) | If(bt) | Loop(bt) => {
                 self.blocks.push(ExprBlock {
                     label: bt.label,
                     pushed_scope: false,
                 });
                 self.resolve_block_type(bt)?;
             }
             TryTable(try_table) => {
-                self.blocks.push(ExprBlock {
-                    label: try_table.block.label,
-                    pushed_scope: false,
-                });
                 self.resolve_block_type(&mut try_table.block)?;
                 for catch in &mut try_table.catches {
                     if let Some(tag) = catch.kind.tag_index_mut() {
                         self.resolver.resolve(tag, Ns::Tag)?;
                     }
                     self.resolve_label(&mut catch.label)?;
                 }
+                self.blocks.push(ExprBlock {
+                    label: try_table.block.label,
+                    pushed_scope: false,
+                });
```

Comment:
cc @eqrion on this change because this will change the text-to-wasm translation for `try_table`. I believe this is fixing a mistake because previously labels reference by name were off-by-one because the `self.blocks` depth was modified before the catch labels were resolved. By modifying them afterwards I think it fixes that.

Reply:
With your change, is it possible for a `catch` to refer to the `try_table` itself? i.e `try_table $me (catch_all $me)`.

My understanding of the instruction would be that this is allowed. But the spec appears to match your implementation here and validate the catch clauses without the try_table added to the labels vector. [1]

Gonna file an upstream issue.

[1] https://webassembly.github.io/exception-handling/core/valid/instructions.html#xref-syntax-instructions-syntax-instr-control-mathsf-try-table-xref-syntax-instructions-syntax-blocktype-mathit-blocktype-xref-syntax-instructions-syntax-catch-mathit-catch-ast-xref-syntax-instructions-syntax-instr-mathit-instr-ast-xref-syntax-instructions-syntax-instr-control-mathsf-end

Reply:
https://github.com/WebAssembly/exception-handling/issues/286

Reply:
Ah yeah I was also suprised as I expected to be able to catch-to-yourself but the spec tests required otherwise which is why I changed this. I'd be happy to update one way or another depending on the resolution of your issue, thanks for filing that!

Reply:
@alexcrichton Reading the room in the issue, not sure it's going to change. So I'd say merging this so it matches the spec is the best way forward.